text/0017-message/0017-message.md

```diff
@@ -0,0 +1,182 @@
+
+To better understand this layered design, consider the internet protocol for example. There is an Ethernet frame, that contains an IP payload. This in turn contains a TCP packet that encapsulates an HTTP payload. Each layer has a certain responsibility, once this responsibility is completed, we move on to the next layer.
+
+The same goes with how we parse messages. The outer layer of the message enables us to map the message to a vertex in the Tangle and perform some basic validation. The next layer may be a transaction that mutates the ledger state. The next layer may provide some extra functionality on the transactions to be used by applications.
+
+By making it possible to add and exchange payloads, we are creating an architecture that can be easily extended to accommodate future needs.
+
+
+# Detailed design
+
+### Data Types
+
+The following are data types that we will use when we specify fields in the message and payloads.
+
+| Name   | Description   |
+| ------ | ------------- |
+| varint | An unsigned numerical value using a dynamic amount of bytes encoded in Little Endian. Evaluate by inspecting the MSB of the first byte. If it is `0` stop. If it is `1` go to the next byte. Continue applying this rule until you reach a byte with MSB `0`. Transform each byte to a 7-bit bit sequence by stripping the MSBs from each byte. Reverse the order of the 7 bit groups (Little Endian). Concatenate all the bits and evaluate as an unsigned integer. [It is similar to how it is done in Protocol Buffers](https://developers.google.com/protocol-buffers/docs/encoding). We must add [validation rules](#Message-Validation) that ensure that the varint is not malleable, and it is limited in size. |
+| uint64  | An unsigned 64 bit integer encoded in Little Endian.
+| ByteArray    | A dynamic size byte array. It is a prefixed by a `varint` that indicates the number of bytes in the array. The following bytes are simply the content of the array.
+| ByteArray[N] | A static size array of size N. Since the size is expected, no need to prefix with a varint   |
+
+
+### Message ID
+The message ID will be the `BLAKE2b-256` hash of the byte contents of the message. It should be used by the nodes to index the messages and by external APIs.
+
+
+### Message Structure
+
+<table>
+    <tr>
+        <th>Name</th>
+        <th>Type</th>
+        <th>Description</th>
+    </tr>
+    <tr>
+        <td>Version</td>
+        <td>varint</td>
+        <td>The message version. The schema specified in this RFC is for version <strong>1</strong> only. </td>
+    </tr>
+    <tr>
+        <td>Parent1 (<code>trunk</code>)</td>
+        <td>ByteArray[32]</td>
+        <td>The Message ID of the first <i>Message</i> we reference.</td>
+    </tr>
+    <tr>
+        <td>Parent2 (<code>branch</code>)</td>
+        <td>ByteArray[32]</td>
+        <td>The Message ID of the second <i>Message</i> we reference.</td>
+    </tr>
+    <tr>
+        <td>Payload Length</td>
+        <td>varint</td>
```

Comment:
What to we gain by using a varint here? Are you planning to have megabyte-sizes messages? Make it a uint32 and reduce complexity instead of making even the header of the message variable in size

Reply:
We wanted to not have too many types in the message.
The idea is that less types means more simple descriptions and code.

Also, for lots of cases where we describe quantities `uint8` or `uint16` could have sufficed.

So the idea is that choosing varints will result mostly in shorter messages

Reply:
I'm not convince it will be simpler code. It's not native primitives in many languages.
So we either implement one ourselves or use the existing one. Both would end up with more validation rules.

Reply:
If `uint8_t` or `uint16_t` is sufficed, why not just use one of them? they are primitive types in most languages. The `varint` is good for data serialization but it is not easy to operate and memory management in a data structure, especially for microcontrollers.

Reply:
due to recent conversations outside of github I am also leaning towards dropping varints

this decision will be finalized on the next protocol team meeting

Reply:
TL;DR: I would rather set fixed reasonable-sized types with some slack (you never know) replacing all varints.

`varints` are somewhat complex types to parse, but I would like to draw your attention to the creation of such message rather than the parsing of it. `varints` type force the entire message structure to "shift" instead of having fixed offsets. This is quite cumbersome when forming a message where all the parts are not necessarily known before-hand, and therefore the byte-requirements of length `varint` types is also unknown: when you fill in the data section of the message, you may see yourself push it forward in order to accommodate the corresponding `varint` size without overflowing into the data section itself. Sure, modern programming languages and libs provide abstract structures for all these, but at a cost of memory reallocation, memory concatenation and other forms of memory juggling. Leveling the participation field ultimately participates in the flexibility of a solution.